### PR TITLE
Revert/return/selfdestruct to properly abort

### PR DESCRIPTION
**Pull request by axic**
Depends on https://github.com/ewasm/design/issues/72 (and https://github.com/ewasm/design/pull/87)

### CODE REVIEW

**Comment by axic** [comment]
@chfast can you review the implementation detail?

File: src/eei.cpp [hunk 1]
```diff
@@ -619,7 +619,7 @@ inline int64_t maxCallGas(int64_t gas) {
 
       result.isRevert = import->base == Name("revert");
 
-      return Literal();
+      throw new EndExecution{};
```
Comment (chfast) [inline src/eei.cpp]: Should be `throw EndExecution{}`. You want to throw an exception of type `EndExecution` not `EndExecution*`.

File: src/hera.cpp [hunk 2]
```diff
@@ -326,7 +326,13 @@ void execute(
 
   Name main = Name("main");
   LiteralList args;
-  instance.callExport(main, args);
+
+  try {
+    instance.callExport(main, args);
+  } catch (EndExecution const&) {
```
Comment (chfast) [inline src/hera.cpp]: This will not work because the exception type is `EndExecution*` (no match). Must be fixed in other place.

**Comment by axic** [comment]
> I'm not a fan of using exceptions for control flow. I don't see the full picture, but I would avoid using C++ exceptions for any eWASM / EVM specified failures.

I don't think there is any other way in binaryen. This is not a failure, but aborting on `return` / `revert`, just like how `exit` works in wasm.

File: src/hera.cpp (hunk 2 above)
Comment (axic) [reply src/hera.cpp]: Fixed both.

**Comment by axic** [comment]
The design change was merged and no test changes are required.

**Pull request merged**